c2rust-analyze/src/main.rs

```diff
@@ -460,6 +460,77 @@ fn label_rvalue_tys<'tcx>(acx: &mut AnalysisCtxt<'_, 'tcx>, mir: &Body<'tcx>) {
     }
 }
 
+fn update_pointer_info<'tcx>(acx: &mut AnalysisCtxt<'_, 'tcx>, mir: &Body<'tcx>) {
+    let mut write_count = HashMap::with_capacity(mir.local_decls.len());
+    let mut rhs_is_ref = HashSet::new();
+
+    for (bb, bb_data) in mir.basic_blocks().iter_enumerated() {
+        for (i, stmt) in bb_data.statements.iter().enumerate() {
+            let (pl, rv) = match &stmt.kind {
+                StatementKind::Assign(x) => &**x,
+                _ => continue,
+            };
+
+            eprintln!(
+                "update_pointer_info: visit assignment: {:?}[{}]: {:?}",
+                bb, i, stmt
+            );
+
+            // Note we ignore `c_void_casts` here.  It shouldn't affect any of the patterns we're
+            // looking for.
+
+            if !pl.is_indirect() {
+                // This is a write directly to `pl.local`.
+                *write_count.entry(pl.local).or_insert(0) += 1;
+                eprintln!("  record write to LHS {:?}", pl);
+            }
+
+            let ref_pl = match *rv {
+                Rvalue::Ref(_rg, _kind, pl) => Some(pl),
+                Rvalue::AddressOf(_mutbl, pl) => Some(pl),
+                _ => None,
+            };
+            if let Some(ref_pl) = ref_pl {
+                // For simplicity, we consider taking the address of a local to be a write.  We
```

Comment:
why is this considered a write? what does that imply downstream?

Reply:
Again, just being conservative.  What we're looking for here is the code that's generated by `&x`/`&mut x` expressions in the source, which consists of a temporary `_1` that's initialized with a statement `_1 = &mut ...;` and which otherwise isn't written in any way.

We might need to do something slightly smarter here in the future, if we eventually need to handle `&mut &mut x`.

Reply:
I think some or all of your response is worth adding as a comment

Reply:
1429293d